### PR TITLE
Bump wallet version to 159900 and remove the `usehd` option

### PR DESCRIPTION
**Pull request by achow101**
Bump the wallet version number to 159900 so that new wallets made without a default key will no longer work on previous versions at all. Also remove the `usehd` option to avoid weird interaction with wallet version numbers and HD-ness of wallets.

### CODE REVIEW

File: src/wallet/wallet.cpp [hunk 1]
```diff
@@ -3849,17 +3845,6 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
 
         walletInstance->SetBestChain(chainActive.GetLocator());
     }
-    else if (gArgs.IsArgSet("-usehd")) {
-        bool useHD = gArgs.GetBoolArg("-usehd", DEFAULT_USE_HD_WALLET);
```
Comment (laanwj) [inline src/wallet/wallet.cpp]: I think we want to keep raising an error if `-usehd` is provided

**Comment by laanwj** [comment]
The option needs to be removed from `GetWalletHelpString` help too

**Comment by MarcoFalke** [comment]
Removal of `usehd` needs a small mention in the release-notes.md

File: src/wallet/wallet.cpp (hunk 1 above)
Comment (achow101) [reply src/wallet/wallet.cpp]: Done

**Comment by laanwj** [comment]
Concept ACK, we need this asap, to prevent problems with wallets created with master on older versions.

File: src/wallet/wallet.cpp [hunk 2]
```diff
@@ -3850,15 +3846,8 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
...
-        if (walletInstance->IsHDEnabled() && !useHD) {
-            InitError(strprintf(_("Error loading %s: You can't disable HD on an already existing HD wallet"), walletFile));
-            return nullptr;
-        }
-        if (!walletInstance->IsHDEnabled() && useHD) {
-            InitError(strprintf(_("Error loading %s: You can't enable HD on an already existing non-HD wallet"), walletFile));
-            return nullptr;
-        }
+        InitError(_("Error: -usehd is no longer a valid option. HD Wallets cannot be disabled."));
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: ultra-nit: usually I think we write "HD wallets".

File: src/wallet/wallet.cpp [hunk 3]
```diff
@@ -3850,15 +3846,8 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
...
-        bool useHD = gArgs.GetBoolArg("-usehd", DEFAULT_USE_HD_WALLET);
-        if (walletInstance->IsHDEnabled() && !useHD) {
-            InitError(strprintf(_("Error loading %s: You can't disable HD on an already existing HD wallet"), walletFile));
-            return nullptr;
-        }
-        if (!walletInstance->IsHDEnabled() && useHD) {
-            InitError(strprintf(_("Error loading %s: You can't enable HD on an already existing non-HD wallet"), walletFile));
-            return nullptr;
-        }
+        InitError(_("Error: -usehd is no longer a valid option. HD wallets cannot be disabled."));
```
Comment (jonasschnelli) [inline src/wallet/wallet.cpp]: nit: maybe we should write "[Creating] HD wallets cannot be disabled". Otherwise it may confuse if you habe usehd=1 (or =0) in your config.

**Comment by jonasschnelli** [comment]
Tested ACK 62fad1266f4ab41453a86ca83141dc1b4791d20f (modulo possible nit)
https://bitcoin.jonasschnelli.ch/build/291

File: src/wallet/wallet.cpp [hunk 4]
```diff
@@ -3829,17 +3829,13 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
 
     if (fFirstRun)
     {
-        // Create new keyUser and set as default key
-        if (gArgs.GetBoolArg("-usehd", DEFAULT_USE_HD_WALLET) && !walletInstance->IsHDEnabled()) {
+        // ensure this wallet.dat can only be opened by clients supporting HD with chain split and has no default key
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: nit: s/has no/expecting no/

File: src/wallet/wallet.cpp [hunk 5]
```diff
@@ -3829,17 +3829,13 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
 
     if (fFirstRun)
     {
-        // Create new keyUser and set as default key
-        if (gArgs.GetBoolArg("-usehd", DEFAULT_USE_HD_WALLET) && !walletInstance->IsHDEnabled()) {
```
Comment (TheBlueMatt) [inline src/wallet/wallet.cpp]: After discussion I believe the removal of the !IsHDEnabled check here is fine but I could totally see it breaking something very subtly, so would be good to get more eyeballs on this line.

**Review by TheBlueMatt — COMMENTED** [review]
utACK 62fad1266f4ab41453a86ca83141dc1b4791d20f

File: src/wallet/wallet.cpp (hunk 5 above)
Comment (jonasschnelli) [reply src/wallet/wallet.cpp]: IMO the `IsHDEnabled()` check was there because you could salvage the wallet which may then would have triggered a `fFirstRun` and recreated a HD master key. But this is no longer true since `fFirstRun` is always false if there is a key in the wallet (and HD master key is also a standard key).

File: src/wallet/wallet.cpp (hunk 3 above)
Comment (sipa) [reply src/wallet/wallet.cpp]: I think the error should only be shown when the value passed to -usehd disagrees with the version number. That way setups with usehd=1 in the config file don't break. The error message could explain how to use -upgradewallet to enable HD on an existing non-HD wallet.

**Comment by achow101** [comment]
Comments addressed. I reverted the `-usehd` warning to just be the original warning with an additional statement for `-usehd=0` option. It now says

> You can't disable HD on an already existing HD wallet or create new non-HD wallets.

**Comment by jonasschnelli** [comment]
utACK 713a92073b3b391f48d4fa146bd334893ec94b8f

File: src/wallet/wallet.cpp [hunk 6]
```diff
@@ -3852,7 +3848,7 @@ CWallet* CWallet::CreateWalletFromFile(const std::string walletFile)
     else if (gArgs.IsArgSet("-usehd")) {
         bool useHD = gArgs.GetBoolArg("-usehd", DEFAULT_USE_HD_WALLET);
         if (walletInstance->IsHDEnabled() && !useHD) {
-            InitError(strprintf(_("Error loading %s: You can't disable HD on an already existing HD wallet"), walletFile));
+            InitError(strprintf(_("Error loading %s: You can't disable HD on an already existing HD wallet or create new non-HD wallets."), walletFile));
```
Comment (MarcoFalke) [inline src/wallet/wallet.cpp]: This new warning will never be printed, since the code branch is neither executed on first start nor later (as the wallet will be hd even when -nousehd)

File: src/wallet/init.cpp [hunk 7]
```diff
@@ -29,7 +29,6 @@ std::string GetWalletHelpString(bool showDebug)
     strUsage += HelpMessageOpt("-salvagewallet", _("Attempt to recover private keys from a corrupt wallet on startup"));
     strUsage += HelpMessageOpt("-spendzeroconfchange", strprintf(_("Spend unconfirmed change when sending transactions (default: %u)"), DEFAULT_SPEND_ZEROCONF_CHANGE));
     strUsage += HelpMessageOpt("-txconfirmtarget=<n>", strprintf(_("If paytxfee is not set, include enough fee so transactions begin confirmation on average within n blocks (default: %u)"), DEFAULT_TX_CONFIRM_TARGET));
-    strUsage += HelpMessageOpt("-usehd", _("Use hierarchical deterministic key generation (HD) after BIP32. Only has effect during wallet creation/first start") + " " + strprintf(_("(default: %u)"), DEFAULT_USE_HD_WALLET));
```
Comment (MarcoFalke) [inline src/wallet/init.cpp]: Needs mention in the release notes and fixup for check-doc.py

File: test/functional/wallet-hd.py [hunk 8]
```diff
@@ -15,17 +15,11 @@ class WalletHDTest(BitcoinTestFramework):
...
         self.num_nodes = 2
-        self.extra_args = [['-usehd=0'], ['-usehd=1', '-keypool=0']]
+        self.extra_args = [[], ['-keypool=0']]
 
     def run_test (self):
         tmpdir = self.options.tmpdir
 
-        # Make sure can't switch off usehd after wallet creation
-        self.stop_node(1)
-        self.assert_start_raises_init_error(1, ['-usehd=0'], 'already existing HD wallet')
```
Comment (MarcoFalke) [inline test/functional/wallet-hd.py]: Why is this test removed?

**Review by maflcko — COMMENTED** [review]
Post merge Concept ACK, but needs feedback addressed

File: test/functional/wallet-hd.py (hunk 8 above)
Comment (achow101) [reply test/functional/wallet-hd.py]: Oops. That was removed originally when I had `-usehd` always return an error regardless.

**Comment by xor-freenet** [comment]
@achow101 
> Also remove the ```usehd``` option to avoid weird interaction with wallet version numbers and HD-ness of wallets.

Can you please clarify what those "weird interactions" are?
Thanks!

**Comment by achow101** [comment]
The version number of a HD wallet is greater than or equal to 130000. But this makes the wallet version number 159900. So, by the previous version number constraint, all newly created wallets must be a HD wallet. Creating a non-HD wallet would mean that the version number has to be less than 130000, but then the new wallet cannot include the fix provided by the new version number. Thus we must either make some hack to allow new non-HD wallets sanely, or don't make new non-HD wallets at all. I chose the latter.

**Comment by MarcoFalke** [comment]
@xor-freenet See #11582

**Comment by xor-freenet** [comment]
@achow101 
> The version number of a HD wallet is greater than or equal to 130000. But this makes the wallet version number 159900. So, by the previous version number constraint, all newly created wallets must be a HD wallet. Creating a non-HD wallet would mean that the version number has to be less than 130000, but then the new wallet cannot include the fix provided by the new version number. Thus we must either make some hack to allow new non-HD wallets sanely, or don't make new non-HD wallets at all. I chose the latter.

Thanks.
This sounds like a job for a simple "boolean isHD" stored somewhere in the wallet for all wallets > version 159900 to track the HD feature independently of the version number.
Can you please add that and re-add the ability to disable HD with ```-usehd=0```?

@MarcoFalke 
> @xor-freenet See #11582

Thanks. I have read it but it doesn't end with the re-addition of ```-usehd=0``` and the reasons given do not justify HD's security drawback of deriving all keys from a single piece of entropy - which is fixed by using non-HD wallets where each address has its own entropy.
That PR was closed because the author needed to fully disable key generation completely and that can be done by other means, so the subject of the security drawback of HD wallets was completely ignored. The PR wasn't really about ```usehd```.

**Comment by xor-freenet** [comment]
@achow101 
(Though I just said that #11582 wasn't really about ```usehd``` please notice I merely meant its unrelated goal of disabling key generation, it still **does** in fact provide code to re-add ```usehd``` which you can re-use.)

**Comment by MarcoFalke** [comment]
If you don't have enough entropy to generate a single seed you surely won't have enough entropy to generate multiple (keypoolsize) keys.

**Comment by achow101** [comment]
@xor-freenet Not to be rude, but if you want that feature, you can either implement it yourself or find someone to implement it for you. Of course just implementing it is not enough and you will need to defend why it is a good idea to allow non-HD wallets and how your implementation does not break backwards compatibility and is safe. I am not interested in doing that.

**Comment by xor-freenet** [comment]
@MarcoFalke 
@achow101 
> If you don't have enough entropy to generate a single seed you surely won't have enough entropy to generate multiple (keypoolsize) keys.

The problem isn't a lack of entropy at key generation, the problem is that the single piece of entropy is going to be re-used for an indefinite amount of years for as long as a wallet is used.
Attacks on the crypto may arise where the attacker can derive the master key from a small set of multiple transactions of the victim wallet (or a bug may just leak the key into a single transaction). Then with HD all coins are gone in the wallet.

In other words: With non-HD wallets address re-use was greatly discouraged by the Bitcoin community because publishing a transaction leaks cryptographic information about the private key. Now with HD it's suddenly OK to re-use something which derives from the same key?

Even worse this also means that cold storage isn't really cold storage because key material is constantly leaked by every transaction you send from it by thumb drive.

**Comment by xor-freenet** [comment]
@achow101 
> @xor-freenet Not to be rude, but if you want that feature, you can either implement it yourself or find someone to implement it for you. Of course just implementing it is not enough and you will need to defend why it is a good idea to allow non-HD wallets and how your implementation does not break backwards compatibility and is safe. I am not interested in doing that.

Oh sorry my problem isn't that I wouldn't understand I should do it if I want it, it's merely that I'm more expertised in Java than C++ and it would take me much more time to implement it than for someone who frequently works with the Bitcoin codebase.
I can certainly shelf out e.g. a $50 bounty (paid in BTC of course) if that motivates you? If not feel free to make up a number :)

W.r.t. to backwards compatibility: Well the removal of ```usehd``` removed a command line option which worked in a previous version and now doesn't work in the current, isn't that the definition of backwards compatibility breakage?

**Comment by MarcoFalke** [comment]
@xor-freenet Are you aware that you can create a wallet with a previous version (create a no-hd wallet) and use it with the current version?

https://github.com/bitcoin/bitcoin/blob/master/doc/release-notes/release-notes-0.16.0.md#hd-wallets-by-default

**Comment by xor-freenet** [comment]
@MarcoFalke 
> @xor-freenet Are you aware that you can create a wallet with a previous version (create a no-hd wallet) and use it with the current version?

Yes, thank you. Requiring the usage of obsolete software versions to access security features is a mutual exclusion though, old software is insecure by definition.

Besides they won't stay available forever, it's entirely possible they stop working on e.g. recent Linux versions due to things changing about Linux. And repairing ```usehd``` won't be easier once the legacy versions have become unusable as the currently existing patches won't apply cleanly anymore due to merge conflicts.

I'd really be very very happy if this issue could be sorted out right away.
I fully accept that the usability benefit of HD wallet backups being permanently valid is good for newbies and thus HD should be the default.
However improved usability for newbies shouldn't hurt the security of professional use and thus it should be possible to disable HD.
Bitcoin Core will have to keep compatibility for non-HD keys forever anyway due to the large amount of existing funds stored in them, so it would be really sad if the feature wasn't accessible for new users anymore due to lack of a < 50 line patch to expose it to the UI.
